scripts/install_backend_python_libs.py

```diff
@@ -29,19 +30,53 @@
 
 
 def _normalize_python_library_name(library_name):
-    """Returns a normalized (lowercase) version of the python library name.
-
-    Python library name strings are case insensitive which means that
-    libraries are equivalent even if the casing of the library names are
-    different. This function normalizes python library names such that
-    identical libraries have the same library name strings.
+    """Returns a normalized version of the python library name.
+
+    Normalization of a library name means converting the library name to
+    lowercase, and removing any "[...]" suffixes that occur. The reason we do
+    this is because of 2 potential confusions when comparing library names that
+    will cause this script to find incorrect mismatches.
+
+    1. Python library name strings are case insensitive which means that
+       libraries are equivalent even if the casing of the library names are
+       different.
+    2. There are certain python libraries with a default version and multiple
+       variants. These variants have names like `library[sub-library]` and
+       signify that it is a version of the library with special support for
+       the sub-library.
+
+    Here are some examples of ambiguities that this function resolves:
+    - 'googleappenginemapreduce' is listed in the 'requirements.txt' file as
+      all lowercase. However, the installed directories has names starting with
+      the string 'GoogleAppEngineMapReduce'. This causes confusion when
+      searching for mismatches because python treats the two library names as
+      different even though they are equivalent.
+    - The name 'google-api-core[grpc]', listed in a 'requirements.txt' file
+      means that a variant of the 'google-api-core' package that supports grpc
+      is required. However, the import names, the package directory names, and
+      the metadata directory names of the installed package do not actually
+      contain the sub-directory identifier. This causes incorrect mismatches to
+      be found because the script treats the installed package's library name,
+      'library', differently from the 'requirements.txt' listed library name,
+      'library[sub-library]'
 
     Args:
         library_name: str. The library name to be normalized.
 
     Returns:
-        str. A normalized library name that is all lowercase.
+        str. A normalized library name.
     """
+    # Remove the special support package designation (e.g [grpc]) in the
+    # brackets when parsing the requirements file to resolve confusion 2 in the
+    # docstring.
+    # NOTE: This does not cause ambiguities because there is no viable scenario
```

Comment:
I think you can add a backend test for this, rather than relying on pip. Have it grab the contents of requirements.in and verify that all line items are unique when normalized.

Reply:
Done, I modified the API a little bit and exposed the normalization functions because this allows us to test with them easier. 
I don't think it makes sense to put them in utils since they are very script-specific(normalization doesn't really mean the same thing in other contexts) and I chose to expose these functions instead of the get_requirements_file_contents() function bc 'requirements.txt' is specific to the current implementation(requirements.txt might not be used in another implementation). However, in other implementations of the installation pipeline, we do need to figure out a way to normalize the directory names and python library names so it makes sense for this to part of the public api.

Reply:
Sounds good. Part of the rationale here should be "we have implemented a backend test {{provide reference}} that verifies that all lines in requirements.txt have a unique normalized name". (I just realized it should probably be both .txt and .in because .txt may have additional deps added and we need to check for no clashes with those too.)

Reply:
Done, added another test for txt. I figured at first that since pip generates from requirements.in, just verifying that requirements.in has distinct libraries is fine but then I realized we also need to verify our assumption that the pip-compile script generates distinct libraries.